engine_details/architecture/object_class.rst

```diff
@@ -244,35 +241,111 @@ Adding signals to a class is done in ``_bind_methods``, using the
 
-References
-----------
+- Objects derived from ``RefCounted`` are reference counted.
+- All other objects are manually memory managed.
+
+The ownership models are fundamentally different. Refer to the section for each respectively to learn how to
+create, store, and free the object.
+
+When you do not know whether an object passed to you (via ``Object *``) is ``RefCounted``, and you need to store it,
+you should store its ``ObjectID`` rather than a pointer (as explained below, in the manual memory management section).
 
-:ref:`RefCounted <class_RefCounted>` inherits from Object and holds a
-reference count. It is the base for reference counted object types.
-Declaring them must be done using Ref<> template. For example:
+When an object is passed to you via :ref:`Variant<class_Variant>`, especially when using deferred callbacks, it is
+possible that the contained ``Object *`` was already freed by the time your function runs.
+Instead of converting directly to ``Object *``, you should use ``get_validated_object``:
 
 .. code-block:: cpp
 
-    class MyReference: public RefCounted {
+    void do_something(Variant p_variant) {
+        Object *object = p_variant.get_validated_object();
+        ERR_FAIL_NULL(object);
+    }
+
+Manual memory management
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Manually memory managed objects are created using ``memnew`` and freed using ``memdelete``:
+
+.. code-block:: cpp
+
+    Node *node = memnew(Node);
+    // ...
+    memdelete(node);
+    node = nullptr;
+
+When you are not the sole owner of an object, storing a pointer to it is dangerous: Any other references to the object
+may call ``memdelete`` at any time, causing your pointer to become a dangling pointer, which will eventually result in a
+crash.
+
+When storing objects you are not the only owner of, you should store its ``ObjectID`` rather than a pointer:
+
+.. code-block:: cpp
+
+    Node *node = memnew(Node);
+    ObjectID node_id = node.get_instance_id();
+    // ...
+    Object *maybe_node = ObjectDB::get_instance(node_id);
+    ERR_FAIL_NULL(maybe_node); // The node may have been freed between calls.
+
+``RefCounted`` memory management
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:ref:`RefCounted <class_RefCounted>` subclasses are memory managed with
+`reference counting semantics <https://en.wikipedia.org/wiki/Reference_counting>`__.
+
+They are constructed using ``memnew``, and should be stored in ``Ref`` instances. When the last ``Ref`` instance is
+dropped, the object automatically self-destructs.
+
+.. code-block:: cpp
+
+    class MyRefCounted: public RefCounted {
         GDCLASS(MyReference, RefCounted);
     };
 
-    Ref<MyReference> myref(memnew(MyReference));
+    Ref<MyRefCounted> my_ref = memnew(MyRefCounted);
+    // ...
+    // Ref holds shared ownership over the object, so the object
+    // will not destruct . As long as you have a valid, non-null
+    // Ref, it can be safely  assumed the object still valid.
```

Comment:
```suggestion
    // will not be freed. As long as you have a valid, non-null
    // Ref, it can be safely assumed the object is still valid.
```
"Destruct" isn't really used elsewhere and introduces a bit of confusing separate terminology, so I'd say to keep it standard